File: 2019/20/20.py
from collections import defaultdict


def parse_file(file):
    grid = {}
    with open(file) as f:
        y = 0
        for line in f.readlines():
            x = 0
            for c in line:
                if c == '\n':
                    continue
                grid[(x, y)] = c
                x += 1
            y += 1

    return grid, x, y


def neighbours(point):
    x, y = point
    return [
        (x + 1, y),
        (x - 1, y),
        (x, y + 1),
        (x, y - 1),
    ]


def inner(point, w, h):
    x, y = point
    return 4 < x < w - 4 and 4 < y < h - 4


def teleports_in(grid):
    teleports = defaultdict(list)
    for coords, v in grid.items():
        if v not in ['A', 'Z', ' ', '.', '#']:
            teleports[v].append(coords)

    out = {}
    for t in teleports.values():
        assert len(t) == 2, t
        out[t[0]] = t[1]
        out[t[1]] = t[0]

    return out


def explore(start, goal, grid, w, h, recurse=False):
    visited = set()
    horizon = [start]
    distance = 0
    portals = teleports_in(grid)
    while horizon:
        new_horizon = []
        distance += 1
        for point in horizon:
            x, y, level = point
            visited.add(point)
            for n in neighbours((x, y)):
                if (*n, level) in visited:
                    continue
                if n not in grid:
                    continue
                if grid[n] == 'Z' and level == 0:
                    return distance
                if grid[n] == '.':
                    new_horizon.append((*n, level))
                if n in portals:
                    if recurse:
                        delta = 1 if inner(n, w, h) else -1
                    else:
                        delta = 0
                    if 0 <= level + delta < 100:
                        new_horizon.append((*portals[n], level + delta))
        horizon = new_horizon
    return distance


def find(start, end, recurse):
    grid, width, height = parse_file('input.map')
    for coords, v in grid.items():
        if v == start:
            current = (*coords, 0)

    return explore(
        current, end, grid, width, height, recurse
    )


print("Part 1:", find('A', 'Z', recurse=False))
print("Part 2:", find('A', 'Z', recurse=True))

File: 2019/20/20.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

type Grid map[Point]string

type Point struct {
	x, y int
}

func parseFile(fileName string) (Grid, Point) {
	data, _ := ioutil.ReadFile(fileName)
	grid := make(Grid)
	y := 0
	x := 0
	width := 0
	height := 0
	for _, line := range strings.Split(string(data), "\n") {
		x = 0
		for _, chr := range strings.Split(line, "") {
			grid[Point{x, y}] = chr
			x++
			width = x
		}
		y++
		height = y
	}
	return grid, Point{width, height}
}

func neighbours(x, y int) [4]Point {
	return [4]Point{
		Point{x + 1, y},
		Point{x - 1, y},
		Point{x, y + 1},
		Point{x, y - 1},
	}
}

func levelDelta(p, dimensions Point, recurse bool) int {
	if !recurse {
		return 0
	}
	xInner := 4 < p.x && p.x < dimensions.x-4
	yInner := 4 < p.y && p.y < dimensions.y-4
	if xInner && yInner {
		return 1
	} else {
		return -1
	}
}

type PortalMap map[Point]Point

func isPortal(c string) bool {
	return !strings.Contains("AZ .#", c)
}

func portalsIn(grid Grid) PortalMap {
	portals := make(map[string][]Point)
	for point, chr := range grid {
		if !isPortal(chr) {
			continue
		}
		if _, ok := portals[chr]; !ok {
			portals[chr] = []Point{}
		}
		portals[chr] = append(portals[chr], point)
	}
	output := make(PortalMap)
	for v, p := range portals {
		if len(p) != 2 {
			fmt.Println("bad len", len(p), v, p)
		}
		output[p[0]] = p[1]
		output[p[1]] = p[0]
	}
	return output
}

type Triple struct {
	x, y, level int
}
type Visited map[Triple]bool

func traverseMaze(recurse bool) int {
	grid, dimensions := parseFile("input.map")
	var start Triple
	for p, v := range grid {
		if v == "A" {
			start = Triple{p.x, p.y, 0}
		}
	}

	visited := make(Visited)
	horizon := []Triple{start}
	portals := portalsIn(grid)

	distance := 0
	for len(horizon) > 0 {
		newHorizon := []Triple{}
		distance++
		for _, p := range horizon {
			visited[p] = true
			for _, n := range neighbours(p.x, p.y) {
				np := Triple{n.x, n.y, p.level}
				if _, ok := visited[np]; ok {
					continue
				}
				if _, ok := grid[n]; !ok {
					continue
				}
				if grid[n] == "Z" && p.level == 0 {
					return distance
				}
				if grid[n] == "." {
					newHorizon = append(newHorizon, np)
				}
				if port, ok := portals[n]; ok {
					delta := levelDelta(n, dimensions, recurse)
					if 0 <= p.level+delta && p.level+delta < 100 {
						nextPoint := Triple{
							port.x,
							port.y,
							p.level + delta}
						newHorizon = append(newHorizon, nextPoint)
					}
				}
			}
		}
		horizon = newHorizon
	}
	return distance
}

func main() {
	fmt.Println("Part 1:", traverseMaze(false))
	fmt.Println("Part 2:", traverseMaze(true))
}
